src/utils/walletConnect.ts

```diff
@@ -261,3 +321,169 @@ export function onSessionProposal(
   );
 }
+
+export async function onSessionRequest(
+  event: SignClientTypes.EventArguments['session_request']
+) {
+  const client = await signClient;
+
+  logger.debug(`WC v2: session_request`, { event });
+
+  const { id, topic } = event;
+  const { method, params } = event.params.request;
+
+  if (isSigningMethod(method)) {
+    // transactions aren't a `[address, message]` tuple
+    const isTransactionMethod = isTransactionDisplayType(method);
+    let { address, message } = parseRPCParams({ method, params });
+    const allWallets = store.getState().wallets.wallets;
+
+    if (!isTransactionMethod) {
+      if (!address || !message) {
+        logger.error(
+          new RainbowError(
+            `WC v2: session_request exited, no address or messsage`
+          ),
+          {
+            address,
+            message,
+          }
+        );
+
+        await client.respond({
+          topic,
+          response: formatJsonRpcError(id, `Invalid RPC params`),
+        });
+
+        return;
+      }
+
+      // for TS only, should never happen
+      if (!allWallets) {
+        logger.error(new RainbowError(`WC v2: allWallets is null`));
+        return;
+      }
+
+      const selectedWallet = findWalletWithAccount(allWallets, address);
+
+      if (!selectedWallet || selectedWallet?.type === WalletTypes.readOnly) {
+        logger.debug(
+          `WC v2: session_request exited, selectedWallet was falsy or read only`
+        );
+
+        await client.respond({
+          topic,
+          response: formatJsonRpcError(id, `Wallet is read-only`),
+        });
+
+        return;
+      }
+    } else {
+      address = params[0].from;
+    }
+
+    const session = client.session.get(topic);
+    const { nativeCurrency, network } = store.getState().settings;
+    const chainId = Number(event.params.chainId.split(':')[1]);
+    const dappNetwork = ethereumUtils.getNetworkFromChainId(chainId);
+    const displayDetails = getRequestDisplayDetails(
+      event.params.request,
+      nativeCurrency,
+      dappNetwork
+    );
+    const peerMeta = session.peer.metadata;
+    const request: RequestData = {
+      clientId: session.topic, // I don't think this is used
+      peerId: session.topic, // I don't think this is used
+      requestId: event.id,
+      dappName:
+        dappNameOverride(peerMeta.name) || peerMeta.name || 'Unknown Dapp',
+      dappScheme: 'unused in WC v2', // only used for deeplinks from WC v1
+      dappUrl: peerMeta.url || 'Unknown URL',
+      displayDetails,
+      imageUrl: maybeSignUri(
+        dappLogoOverride(peerMeta.url) || peerMeta.icons[0]
+      ),
+      payload: event.params.request,
+      walletConnectV2RequestValues: {
+        sessionRequestEvent: event,
+        // @ts-ignore we assign address above
+        address, // required by screen
+        chainId, // required by screen
+      },
+    };
+
+    logger.debug(`request`, { request });
```

Comment:
Maybe it makes sense to add a separate debug context for Wallet Connect?

Reply:
I'll go back through and add this as a fast-follow later this week before release 👍